openwebwork/webwork2
# Feedback form expansion - REVISED

# Purpose

Our institution uses the feedbackMacro_form function to direct students to ask questions on a Q&A site (https://openlab.citytech.cuny.edu/ol-webwork) where their questions are aggregated based on having the "same" problem (same pg file, different seeds). That way students have an FAQ on a per-problem basis.

We need some additional info from WeBWorK because the Q&A site does not share accounts with WeBWorK.

# Description

Okay, so - as suggested ( #957 ) - I restructured my expansion of (the necessary, for our Q&A WordPress plugin) form data sent through the `feedbackMacro_form` function. 

1. I removed my changes from `ContentGenerator.pm` and instead expanded the parameters passed to `feedbackMacro` when it is called from `output_email_instructor` in `CG\Problem.pm` and `output_footer` in `CG\ProblemUtil\ProblemUtil.pm`

2. In the process, I found multiple clusters of repeated code - so I shifted them into their own subroutines in WeBWorK::Util. The repositioned code now comprises two new subroutines: `fetchEmailRecipients` and `generateURLs` (which are made available through ContentGenerator.pm). These new functions helped to simplify the process of adding to the form data sent through the feedbackMacro

3. Please forgive the somewhat-large diffs on ContentGenerator.pm and Problem.pm - I have been using Atom as my editor, and it auto-strips trailing whitespace. The **only** relevant change to ContentGenerator.pm is line 62 (loading the new functions from Util); and in Problem.pm, the only meaningful change is to `output_email_instructor` starting on line 2116.

# Testing 

1. One should start by testing that their usual email functions are not affected by these changes.
1. As a student, use the email instructor button and confirm that it works as expected
1. Load up a JITAR set (make sure that there are a limited number of attempts!) - and, as a student, trigger the jitar-warning email by exhausting your allotted attempts. Confirm that an email is sent to the instructor (test TA account too?) warning that a student failed to complete a JITAR set.
1. add the following to the course.conf: 
`$courseURLs{feedbackFormURL} = 'http://localhost/test.php';`
`$feedback_button_name = "Test the PR"`
1. create /var/www/html/test.php with the contents:
```
<html>
<head><title>PHP Get Results</title></head>

<body>

<?php

// Show all form data submitted
foreach($_POST as $key=>$value){
    echo $key, ' => ', htmlspecialchars($value), "<br/>";
}

?>

</body>
</html>
```
1. Visit a problem page and click the (used to be email instructor) button that says "Test the PR"; confirm that a new page opens with a bunch of key => values

# New data provided 

* emailableURL => http://localhost/webwork2/CourseID/problemSetName/problemNumber/?showOldAnswers=1&effectiveUser=student&displayMode=MathJax
    (this was the only thing instructors missed about the email instructor button - we bring it along so that the Q&A site can include it with the notification email) 

* randomSeed => 2013
    (not used by our Q&A site yet, but may be useful in re-rendering the student's problem -- XMLRPC -- instead of decoding the pg64 string that gets sent) 

* notifyAddresses => `Administrator <fake@email.com>;"Prof. Lastname" <lastname@uni.edu>`
    (all users who have permission to receive feedback emails - now our Q&A site is informed of who should receive notification that this user asked a question...)

* problemPath => local/folder/some-problem.pg 
    (this is our unique ID for problems. It maintains a consistent basis for identifying "identical" problems when instructors can remove or reorder problems...)

* studentName => active user's Full Name
    (because our Q&A site does not share single sign-on with WeBWorK - so we can send notifications to instructors/TAs and they 100% know who the student is)
